This work was performed on Arch Linux 6.17.
Requires the gap, sympy and sage packages to be installed.

# **EASY LEVEL**

---

**Task 1**

In [1]:
def is_palindrome(n: int) -> bool:
    s = str(n)
    return s == s[::-1]

def sieve(limit: int) -> list[bool]: # Sieve of Eratosthenes
    is_prime = [True] * limit
    is_prime[0] = is_prime[1] = False

    for i in range(2, int(limit ** 0.5) + 1):
        if is_prime[i]:
            for j in range(i*i, limit, i):
                is_prime[j] = False
    return is_prime

def get_cyclic_permutations(n: int) -> list[int]:
    s = str(n)
    return [int(s[i:] + s[:i]) for i in range(len(s))]

def palindrome_squares_and_circular_primes() -> tuple[list[int], list[int]]:
    # Part 1
    palindromes = []

    for a in range(1, 10**5):
        if is_palindrome(a) and is_palindrome(a * a):
            palindromes.append(a)

    # Part 2
    limit = 10**6
    is_prime = sieve(limit)
    circular_primes = set()

    for p in range(2, limit):
        if not is_prime[p]:
            continue

        digits = str(p)
        if '0' in digits: # Avoiding zeros in cyclic permutations
            continue

        permutations = get_cyclic_permutations(p)

        if all(is_prime[perm] for perm in permutations):
            circular_primes.update(perm for perm in permutations if perm < limit)

    circular_primes = sorted(circular_primes)
    return palindromes, circular_primes

In [2]:
palindrome_squares_and_circular_primes()

([1,
  2,
  3,
  11,
  22,
  101,
  111,
  121,
  202,
  212,
  1001,
  1111,
  2002,
  10001,
  10101,
  10201,
  11011,
  11111,
  11211,
  20002,
  20102],
 [2,
  3,
  5,
  7,
  11,
  13,
  17,
  31,
  37,
  71,
  73,
  79,
  97,
  113,
  131,
  197,
  199,
  311,
  337,
  373,
  719,
  733,
  919,
  971,
  991,
  1193,
  1931,
  3119,
  3779,
  7793,
  7937,
  9311,
  9377,
  11939,
  19391,
  19937,
  37199,
  39119,
  71993,
  91193,
  93719,
  93911,
  99371,
  193939,
  199933,
  319993,
  331999,
  391939,
  393919,
  919393,
  933199,
  939193,
  939391,
  993319,
  999331])

---

**Task 2**

In [3]:
def is_prime(n: int) -> bool:
    if n < 2:
        return False

    if n == 2:
        return True

    if n % 2 == 0:
        return False

    i = 3
    while i * i <= n:
        if n % i == 0:
            return False
        i += 2
    return True

def generate_numbers_with_digits(digit1: int, digit2: int, count: int) -> list[int]:
    primes = []
    current_level = [digit1, digit2]

    while len(primes) < count:
        next_level = []

        for num in current_level:
            if is_prime(num):
                primes.append(num)

                if len(primes) == count:
                    return primes

            next_level.append(num * 10 + digit1)
            next_level.append(num * 10 + digit2)

        current_level = next_level

    return primes

def primes_with_two_digits() -> dict[str, list[int]]:
    result = {}
    pairs = [('13', 1, 3), ('15', 1, 5), ('17', 1, 7), ('19', 1, 9)]

    for key, d1, d2 in pairs:
        primes = generate_numbers_with_digits(d1, d2, 100)
        result[key] = primes

    return result

In [4]:
primes_dict = primes_with_two_digits()
for key, primes in primes_dict.items():
  print(f"'{key}': {primes},")

'13': [3, 11, 13, 31, 113, 131, 311, 313, 331, 3313, 3331, 11113, 11131, 11311, 13313, 13331, 31333, 33113, 33311, 33331, 113111, 113131, 131111, 131113, 131311, 311111, 313133, 313331, 313333, 331333, 333131, 333331, 1111333, 1131113, 1131131, 1131133, 1131331, 1133131, 1133333, 1311131, 1311311, 1313311, 1331333, 1333133, 1333313, 1333331, 3111131, 3111313, 3111331, 3113111, 3113333, 3131113, 3131311, 3133111, 3133331, 3311131, 3331331, 3331333, 3333131, 3333133, 3333311, 3333313, 3333331, 11111131, 11111311, 11113111, 11131111, 11311133, 11313311, 11313331, 11331311, 11333111, 11333131, 13111333, 13131133, 13131331, 13133111, 13133311, 13311113, 13311313, 31111313, 31113113, 31113311, 31133131, 31311113, 31311131, 31333333, 33111311, 33113111, 33113131, 33313333, 33333133, 33333331, 111111113, 111111131, 111111313, 111111331, 111113111, 111113113, 111113131],
'15': [5, 11, 151, 1151, 1511, 11551, 15511, 15551, 51151, 51511, 51551, 55511, 115151, 511111, 511151, 515111, 1111151, 1115

---

**Task 3**

In [5]:
LIMIT_PAIRS = 1000

def is_prime(n: int) -> bool:
    if n < 2:
        return False

    if n == 2:
        return True

    if n % 2 == 0:
        return False

    i = 3

    while i * i <= n:
        if n % i == 0:
            return False
        i += 2
    return True

def twin_primes_analysis(limit_pairs: int = LIMIT_PAIRS) -> tuple[list[tuple[int, int]], list[float]]:
    primes = []
    twin_primes = []
    ratios = []

    current = 2

    while len(twin_primes) < limit_pairs:
        if is_prime(current):
            primes.append(current)

        if len(primes) >= 2 and current - primes[-2] == 2:
            twin_primes.append((primes[-2], current))
            ratio = len(twin_primes) / len(primes)
            ratios.append(ratio)

        current += 1

    return twin_primes, ratios

In [6]:
twin_primes, ratios = twin_primes_analysis()

print("The first 10 pairs of twins:")
for i, pair in enumerate(twin_primes[:10]):
  print(f"{i+1}: {pair}")

print("\nThe ratio of the number of pairs of twins to the total number of prime numbers:")
print("First 10 values:", [f"{r:.3f}" for r in ratios[:10]])
print("Last 10 values:", [f"{r:.3f}" for r in ratios[-10:]])

The first 10 pairs of twins:
1: (2, 4)
2: (3, 5)
3: (5, 7)
4: (11, 13)
5: (17, 19)
6: (29, 31)
7: (41, 43)
8: (59, 61)
9: (71, 73)
10: (101, 103)

The ratio of the number of pairs of twins to the total number of prime numbers:
First 10 values: ['0.500', '0.667', '0.750', '0.667', '0.625', '0.545', '0.500', '0.444', '0.429', '0.370']
Last 10 values: ['0.129', '0.129', '0.129', '0.129', '0.129', '0.129', '0.129', '0.129', '0.129', '0.129']


Note that according to the theory of distribution of prime numbers, the number of prime numbers is approximately equal to $\frac{n}{\ln n}$, where $n$ - number of ordinary numbers, means that: $\frac{1}{\ln n}$ - ratio of the number of prime numbers to ordinary numbers.


---


According to the Harley-Littlewood hypothesis, the number of pairs of simple twins is: $C * \frac{n}{(\ln n)^2}, where\ C ≈ 0.66016 - const$. Then the ratio of the number of pairs of prime twins to the number of prime numbers is equal to: $C * \frac{1}{\ln n}$. While testing this formula I came to the conclusion that this constant starts to work with $n > 10^7$, and for our task we can remove this constant. Then the number of simple twin pairs when $n < 10^7$ approximately equal to: $\frac{1}{\ln n}$. Below I have given an example of my testing function for $n = 1000$.

---

Link to the material where I read about Harley-Littlewood hypothesis - https://en.wikipedia.org/wiki/First_Hardy%E2%80%93Littlewood_conjecture

In [7]:
import math
n = 1000
freq = (1 / math.log(n))
f'{freq:.3f}'

'0.145'

Note that the formula gives approximate values.

Thus, the ratio of the number of pairs of twins to the total number of prime numbers:

---
$when\ n = 1000: $
1.   By code: 0.129
2.   By derived formula: 0.145
---
$when\ n = 10000: $

1.   By code: 0.103
2.   By derived formula: 0.109
---
$when\ n = 50000: $

1.   By code: 0.090
2.   By derived formula: 0.092
---
Also note that $\frac{1}{\ln n} \to 0 \quad when\ n \to \infty$.

---
---

**Task 4**

In [8]:
import math
from sympy import factorint

def factorial_plus_one_factors() -> dict[int, dict[int, int]]:
    result = {}
    for n in range(2, 51):
        num = math.factorial(n) + 1
        factors = factorint(num)
        result[n] = factors
    return result

In [9]:
result = factorial_plus_one_factors()

max_diff_primes = 0
max_n = 0
large_prime_cases = []

for n, factors in result.items():
    num_distinct_primes = len(factors)
    if num_distinct_primes > max_diff_primes:
        max_diff_primes = num_distinct_primes
        max_n = n
    if any(p > 10**6 for p in factors):
        large_prime_cases.append(n)

print(result)
print('')
print(f"\nMaximum number of distinct prime divisors: {max_diff_primes} (when n={max_n})")
print(f"Cases with prime factors > 10^6: {large_prime_cases}")

{2: {3: 1}, 3: {7: 1}, 4: {5: 2}, 5: {11: 2}, 6: {7: 1, 103: 1}, 7: {71: 2}, 8: {61: 1, 661: 1}, 9: {19: 1, 71: 1, 269: 1}, 10: {11: 1, 329891: 1}, 11: {39916801: 1}, 12: {13: 2, 2834329: 1}, 13: {83: 1, 75024347: 1}, 14: {23: 1, 3790360487: 1}, 15: {59: 1, 479: 1, 46271341: 1}, 16: {17: 1, 61: 1, 137: 1, 139: 1, 1059511: 1}, 17: {661: 1, 537913: 1, 1000357: 1}, 18: {19: 1, 23: 1, 29: 1, 61: 1, 67: 1, 123610951: 1}, 19: {71: 1, 1713311273363831: 1}, 20: {20639383: 1, 117876683047: 1}, 21: {43: 1, 439429: 1, 2703875815783: 1}, 22: {23: 1, 521: 1, 93799610095769647: 1}, 23: {47: 2, 79: 1, 148139754736864591: 1}, 24: {811: 1, 765041185860961084291: 1}, 25: {401: 1, 38681321803817920159601: 1}, 26: {1697: 1, 237649652991517758152033: 1}, 27: {10888869450418352160768000001: 1}, 28: {29: 1, 10513391193507374500051862069: 1}, 29: {14557: 1, 218568437: 1, 2778942057555023489: 1}, 30: {31: 1, 12421: 1, 82561: 1, 1080941: 1, 7719068319927551: 1}, 31: {257: 1, 95101: 1, 3038779: 1, 11071448528130

---

**Task 5**

In [10]:
import time
from sympy import factorint, totient

def gcd(a: int, b: int) -> int:
    while b:
        a, b = b, a % b
    return a

def euler_phi_direct(n: int) -> int:
    count = 0
    for k in range(1, n + 1):
        if gcd(k, n) == 1:
            count += 1
    return count

def euler_phi_factor(n: int) -> int:
    if n == 1:
        return 1

    result = n
    factors = factorint(n)

    for p in factors:
        result *= (p - 1)
        result //= p

    return result

def compare_euler_phi_methods(test_values: list[int]) -> dict[str, list[float]]:
    times_direct = []
    times_factor = []
    times_builtin = []

    for n in test_values:
        # Method 1: Brute Force
        start = time.time()
        result1 = euler_phi_direct(n)
        times_direct.append(time.time() - start)

        # Method 2: Factorization
        start = time.time()
        result2 = euler_phi_factor(n)
        times_factor.append(time.time() - start)

        # Method 3: Built-in Function
        start = time.time()
        result3 = totient(n)
        times_builtin.append(time.time() - start)

        if result1 != result2 or result2 != result3:
            print(f"Error: different results for n={n}: {result1}, {result2}, {result3}")
            raise ValueError

    return {
        'direct': times_direct,
        'factor': times_factor,
        'builtin': times_builtin
    }

In [11]:
test_values = [100000, 500000, 1000000, 5000000]

print("Comparison of methods for calculating the Euler function:")
print("n\t\tBrute Force\tFactorization\tBuilt-in")
print("-" * 60)

times = compare_euler_phi_methods(test_values)

for i, n in enumerate(test_values):
    print(f"{n:8d}\t{times['direct'][i]:.6f}\t{times['factor'][i]:.6f}\t{times['builtin'][i]:.6f}")

Comparison of methods for calculating the Euler function:
n		Brute Force	Factorization	Built-in
------------------------------------------------------------
  100000	0.121054	0.000086	0.000329
  500000	1.189629	0.000080	0.000372
 1000000	1.554773	0.000076	0.000361
 5000000	7.392010	0.000051	0.000215




---
---


# **NORMAL LEVEL**
---

In [12]:
ISU = 501773
N = 13 # ISU % 20

m = 7 # 4 + (N % 5)
n = 5 # 2 + (N % 10)
k = 7 # 1 + (N % 7)
n1 = 1 # N % 6
n2 = 2 # (N + 1) % 6
n3 = 3 # (N + 2) % 6

p = 23
s = 17
r = 45
t = 12

---

**Task 1**

For this task I used the GAP script (normal_q1.gap), which can be run with the command *gap -q normal_q1.gap*, the output file is *normal_q1_output.txt*

**total_subgroups** = 11300

**Random Subgroup**
- **index**: 9014
- **order**: 24 (size of subgroup)  
- **generators**: [(1,7,6), (5,7,6), (3,4)] (3 elements that create the subgroup)

**Fixed Subgroup (index 14)**
- **order**: 2 (only 2 elements)
- **generator**: [(2,5)] (single transposition)
- **index in S₇**: 2520 (5040/2 = 2520 cosets)
- **normal**: false

**Cosets**
- **left_cosets**: 2520 (gH = {g, g(2,5)})
- **right_cosets**: 2520 (Hg = {g, (2,5)g})

**Example difference**:  
For g=(5,6): left coset has (2,5,6), right coset has (2,6,5)  
This proves the subgroup is not normal.

In [13]:
def subgroups_of_Sm(N: int) -> dict:
    # Parsing
    data = {}
    with open('normal_q1_output.txt', 'r') as f:
        for line in f:
            line = line.strip()
            if '=' in line:
                key, value = line.split('=', 1)
                key = key.strip()
                value = value.strip()
                
                if key == 'total_subgroups':
                    data[key] = int(value)
                elif key in ['random_index', 'random_order', 'fixed_index', 'fixed_order', 
                            'fixed_index_in_G', 'left_cosets_count', 'right_cosets_count']:
                    data[key] = int(value)
                elif key == 'fixed_normal':
                    data[key] = (value.lower() == 'true')
                elif key in ['random_generators', 'fixed_generators', 
                            'left_cosets_example', 'right_cosets_example']:
                    data[key] = value
    return {
        "m": m,
        "total_subgroups": data['total_subgroups'],
        
        "random_subgroup": {
            "index": data['random_index'],
            "order": data['random_order'],
            "generators": data['random_generators']
        },
        
        "fixed_subgroup": {
            "index": data['fixed_index'],
            "order": data['fixed_order'],
            "generators": data['fixed_generators'],
            "group_index": data['fixed_index_in_G'],
            "is_normal": data['fixed_normal'],
            "left_cosets_count": data['left_cosets_count'],
            "right_cosets_count": data['right_cosets_count'],
            "left_cosets_example": data['left_cosets_example'],
            "right_cosets_example": data['right_cosets_example']
        }
    }

In [14]:
result = subgroups_of_Sm(N)
print("Total subgroups:", result["total_subgroups"])
print("Random subgroup:", result["random_subgroup"])
print("Fixed subgroup:", result["fixed_subgroup"])

Total subgroups: 11300
Random subgroup: {'index': 9014, 'order': 24, 'generators': '[ (1,7,6), (5,7,6), (3,4) ]'}
Fixed subgroup: {'index': 14, 'order': 2, 'generators': '[ (2,5) ]', 'group_index': 2520, 'is_normal': False, 'left_cosets_count': 2520, 'right_cosets_count': 2520, 'left_cosets_example': '[[ (), (2,5) ], [ (6,7), (2,5)(6,7) ], [ (5,6), (2,5,6) ]]', 'right_cosets_example': '[[ (), (2,5) ], [ (6,7), (2,5)(6,7) ], [ (5,6), (2,6,5) ]]'}


---

**Task 2**

For this task I also used the GAP script (normal_q2.gap), which can be run with the command *gap -q normal_q2.gap*, the output file is *normal_q2_output.txt*

**Data:**

**G**
- **g**: (4,6,7,5)
- **order_g**: 4

**G1**
- **g1**: (4,6,7,5)
- **order_g1**: 4

**G2**
- **g2**: (4,7)(5,6)
- **order_g2**: 2

**G3**
- **g3**: (4,5,7,6)
- **order_g3**: 4

In [15]:
def element_powers_in_Sm(N: int) -> dict:
    # Parsing
    data = {}
    with open('normal_q2.txt', 'r') as f:
        for line in f:
            line = line.strip()
            if '=' in line:
                key, value = line.split('=', 1)
                key = key.strip()
                value = value.strip()
                
                if key.startswith('order'):
                    data[key] = int(value)
                else:
                    data[key] = value
    
    return {
        'n1': n1,
        'n2': n2, 
        'n3': n3,
        'g': data['g'],
        'order_g': data['order_g'],
        'g1': data['g1'],
        'order_g1': data['order_g1'],
        'cyclic_subgroup_order1': data['order_g1'],
        'g2': data['g2'],
        'order_g2': data['order_g2'],
        'cyclic_subgroup_order2': data['order_g2'],
        'g3': data['g3'],
        'order_g3': data['order_g3'],
        'cyclic_subgroup_order3': data['order_g3']
    }

In [16]:
element_powers_in_Sm(N)

{'n1': 1,
 'n2': 2,
 'n3': 3,
 'g': '(4,6,7,5)',
 'order_g': 4,
 'g1': '(4,6,7,5)',
 'order_g1': 4,
 'cyclic_subgroup_order1': 4,
 'g2': '(4,7)(5,6)',
 'order_g2': 2,
 'cyclic_subgroup_order2': 2,
 'g3': '(4,5,7,6)',
 'order_g3': 4,
 'cyclic_subgroup_order3': 4}

**Results:**

**Data**
- **g = (4,6,7,5)** - 4-cycle (order 4)
- **n₁=1, n₂=2, n₃=3**

**g1 = (4,6,7,5)**
- **Element order**: 4 (4-cycle)
- **Subgroup order**: 4 (<g1\> = {e, g, g2, g3})

**g2 = (4,7)(5,6)**
- **Element order**: 2 (product of two transpositions)
- **Subgroup order**: 2 (<g2\> = {e, g2})

**g3 = (4,5,7,6)**
- **Element order**: 4 (inverse element of g in 4-cycle)
- **Subgroup order**: 4 (<g3\> = <g\>)

---

**Task 3**

For this task I also used the GAP script (normal_q3.gap), which can be run with the command *gap -q normal_q3.gap*, the output file is *normal_q3_output.txt*

**Data**
- **N**: 13
- **m**: 7
- **n**: 5

In [17]:
def solve_sigma_power_eq(N: int) -> dict:
    # Parsing
    data = {}
    with open('normal_q3_output.txt', 'r') as f:
        for line in f:
            line = line.strip()
            if '=' in line:
                key, value = line.split('=', 1)
                key = key.strip()
                value = value.strip()
                
                if key == 'total_solutions':
                    data[key] = int(value)
                elif key == 'solution':
                    data[key] = value.strip('()')
    
    return {
        'equation': f'sigma^{n} = (1 2 3 ... {m - 1})',
        'total_solutions': data['total_solutions'],
        'solution': data['solution']
    }

In [18]:
solve_sigma_power_eq(N)

{'equation': 'sigma^5 = (1 2 3 ... 6)',
 'total_solutions': 1,
 'solution': '1,6,5,4,3,2'}

**Result:**
- **total_solutions**: 1
- **solution**= (1,6,5,4,3,2)

Note that $\sigma$ is a 6-cycle:
- $\sigma^6 = e$
- $\sigma^5 = \sigma^{-1}(since\ \sigma^5 = \sigma^{-1} * \sigma^6 = \sigma^{-1})$
- $\sigma^{-1} = (1 2 3 4 5 6)$
- Therefore: $\sigma = (1 2 3 4 5 6)^{-1} = (1 6 5 4 3 2)$, and there is only one solution. 

---

**Task 4**

For this task I also used the GAP script (normal_q4.gap), which can be run with the command *gap -q normal_q4.gap*, the output file is *normal_q4_output.txt*

**Data**
- **m** = 7
- **k** = 7

In [19]:
def elements_of_order_k_in_cyclic_group(N: int) -> dict:
    # Parsing
    data = {}
    with open('normal_q4_output.txt', 'r') as f:
        for line in f:
            line = line.strip()
            if '=' in line:
                key, value = line.split('=', 1)
                key = key.strip()
                value = value.strip()
                
                if key == 'm':
                    data[key] = int(value)
                elif key == 'k':
                    data[key] = int(value)
                elif key == 'elements_satisfying':
                    data[key] = value
                elif key == 'elements_of_order_k':
                    data[key] = value
    
    return {
        'elements_satisfying_g^k=e': data['elements_satisfying'],
        'elements_of_order_k': data['elements_of_order_k']
    }

In [20]:
elements_of_order_k_in_cyclic_group(N)

{'elements_satisfying_g^k=e': '[ <identity> of ..., f1, f1^2, f1^3, f1^4, f1^5, f1^6 ]',
 'elements_of_order_k': '[ f1, f1^2, f1^3, f1^4, f1^5, f1^6 ]'}

**Results:**
- **Elements satisfying $g^7 = e$**: all 7 elements
- **Elements of order 7**: all 6 non-identity elements

**Why**:

In cyclic group of order k:
- By Lagrange's theorem: $g^k = e$ for all $g$
- All non-identity elements have order $k$
- Number of elements of order $k = k - 1$

---

**Test 5**

For this task I also used the GAP script (normal_q5.gap), which can be run with the command *gap -q normal_q5.gap*, the output file is *normal_q5_output.txt*

**Data**
- **m** = 7

In [21]:
def subgroups_of_Zm_star(N: int) -> list:
    # Parsing
    data = {}
    with open('normal_q5_output.txt', 'r') as f:
        for line in f:
            line = line.strip()
            if '=' in line:
                key, value = line.split('=', 1)
                key = key.strip()
                value = value.strip()
                
                if key == 'm':
                    data[key] = int(value)
                elif key == 'subgroups':
                    subgroups = value.strip('[]').split('", "')
                    subgroups = [s.strip('"') for s in subgroups]
                    data[key] = subgroups
    
    return {
        'subgroups': data['subgroups']
    }

In [23]:
subgroups_of_Zm_star(N)

{'subgroups': [' Group( Z(7)^0 ), Group( [ Z(7)^3 ] ), Group( [ Z(7)^2 ] ), Group( [ Z(7)^3, Z(7)^2 ] ) ']}

**Results:**
- **subgroups**: ['Group(Z(7)^0), Group([Z(7)^3]), Group([Z(7)^2]),']
---
- Group(Z(7)^0) = {1}
- Group([Z(7)^3]) = {1, 6}
- Group([Z(7)^2]) = {1, 2, 4}
- Group([Z(7)^3,Z(7)^2]) = {1, 2, 3, 4, 5, 6}


---

**Task 6**

For this task I also used the GAP script (normal_q6.gap), which can be run with the command *gap -q normal_q6.gap*, the output file is *normal_q6_output.txt*

**Data**
- **N**: 13
- **p**: 23
- **r**: 45
- **s**: 17

In [24]:
def order_of_sr(N: int) -> int:
    # Parsing
    data = {}
    with open('normal_q6_output.txt', 'r') as f:
        for line in f:
            line = line.strip()
            if '=' in line:
                key, value = line.split('=', 1)
                key = key.strip()
                value = value.strip()
                
                if key == 'order':
                    data[key] = int(value)
    
    return data['order']

In [25]:
order_of_sr(N)

22

**Results:**
- **order**: 22

---

**Task 7**

For this task I also used the GAP script (normal_q7.gap), which can be run with the command *gap -q normal_q7.gap*, the output file is *normal_q7_output.txt*

**Data**
- **N**: 13
- **p**: 23
- **t**: 12

In [26]:
def order_and_primitivity_of_t(N: int) -> dict:
    # Parsing
    data = {}
    with open('normal_q7_output.txt', 'r') as f:
        for line in f:
            line = line.strip()
            if '=' in line:
                key, value = line.split('=', 1)
                key = key.strip()
                value = value.strip()
                
                if key == 'order':
                    data[key] = int(value)
                elif key == 'is_primitive':
                    data[key] = (value.lower() == 'true')
    
    return {
        'order': data['order'],
        'is_primitive': data['is_primitive']
    }

In [27]:
order_and_primitivity_of_t(N)

{'order': 11, 'is_primitive': False}

**Results:**
- **order**: 11
- **is_primitive**: False

**Why:**
The element $t = 12$ has order $11$ in $Z_{23}^*$, which is less than $22$, so it is not a primitive root

---

**Task 8**

For this task I also used the GAP script (normal_q8.gap), which can be run with the command *gap -q normal_q8.gap*, the output file is *normal_q8_output.txt*

**Data**
- **N**: 13
- **m**: 7

In [32]:
def generators_of_Zm_star(N: int) -> list:
    # Parsing
    data = {}
    with open('normal_q8_output.txt', 'r') as f:
        for line in f:
            line = line.strip()
            if '=' in line:
                key, value = line.split('=', 1)
                key = key.strip()
                value = value.strip()
                
                if key == 'm':
                    data[key] = int(value)
                elif key == 'generators':
                    generators = value.strip('[]').split(',')
                    generators = [int(g.strip()) for g in generators if g.strip()]
                    data[key] = generators
    
    return {
        'generators': data['generators']
    }

In [33]:
generators_of_Zm_star(N)

{'generators': [3, 5]}

**Results:**
- **generators** = [3, 5]

**Why:**
- $3^1=3, 3^2=2, 3^3=6, 3^4=4, 3^5=5, 3^6=1$ $\Rightarrow$ order $6$
- $5^1=5, 5^2=4, 5^3=6, 5^4=2, 5^5=3, 5^6=1$ $\Rightarrow$ order $6$

---

**Task 9**

For this task I also used the GAP script (normal_q9.gap), which can be run with the command *gap -q normal_q9.gap*, the output file is *normal_q9_output.txt*

**Data**
- **N**: 13
- **m**: 7
- **t**: 12

In [45]:
def cyclic_subgroup_in_Zm_additive(N: int) -> dict:
    # Parsing
    data = {}
    with open('normal_q9_output.txt', 'r') as f:
        for line in f:
            line = line.strip()
            if '=' in line:
                key, value = line.split('=', 1)
                key = key.strip()
                value = value.strip()
                
                if key == 'm':
                    data[key] = int(value)
                elif key == 't':
                    data[key] = int(value)
                elif key == 'subgroup_order':
                    data[key] = int(value)
                elif key == 'subgroup_elements':
                    elements = value.strip('[]').split(',')
                    elements = [int(e.strip()) for e in elements if e.strip()]
                    data[key] = elements
                elif key == 'generators':
                    generators = value.strip('[]').split(',')
                    generators = [int(g.strip()) for g in generators if g.strip()]
                    data[key] = generators
    
    return {
        'subgroup_order': data['subgroup_order'],
        'subgroup_elements': data['subgroup_elements'],
        'generators': data['generators']
    }

In [43]:
cyclic_subgroup_in_Zm_additive(N)

{'subgroup_order': 7,
 'subgroup_elements': [0, 1, 2, 3, 4, 5, 6],
 'generators': [1, 2, 3, 4, 5, 6]}

**Result:**
- **subgroup_order**: 7
- **subgroup_elements**: [0, 1, 2, 3, 4, 5, 6]
- **generators**: [1, 2, 3, 4, 5, 6]

**Relation with t:**
All generators have the same $gcd$ with $m$ as $t$
- $gcd(1,7)=1, gcd(2,7)=1, ..., gcd(6,7)=1$
- $gcd(5,7)=1$

---

**Task 10**

For this task I also used the GAP script (normal_q10.gap), which can be run with the command *gap -q normal_q10.gap*, the output file is *normal_q10_output.txt*

**Data**
- **N**: 13
- **m**: 7
- **t**: 12

In [46]:
def isomorphism_of_cyclic_subgroup_Zm_star(N: int) -> dict:
    # Parsing
    data = {}
    with open('normal_q10_output.txt', 'r') as f:
        for line in f:
            line = line.strip()
            if '=' in line:
                key, value = line.split('=', 1)
                key = key.strip()
                value = value.strip()
                
                if key == 'm':
                    data[key] = int(value)
                elif key == 't':
                    data[key] = int(value)
                elif key == 'subgroup_order':
                    data[key] = int(value)
                elif key == 'isomorphic_to_Sd':
                    data[key] = value
    
    return {
        'subgroup_order': data['subgroup_order'],
        'isomorphic_to_Sd': data['isomorphic_to_Sd']
    }

In [ ]:
isomorphism_of_cyclic_subgroup_Zm_star(N)

{'subgroup_order': 6,
 'isomorphic_to_Sd': 'cyclic subgroup of S_6 generated by 6-cycle'}

**Result:**
- **subgroup_order**: 6
- **isomorphic_to_Sd**: cyclic subgroup of $S_6$ generated by 6-cycle

**The subgroup $(5)$ of $Z_7$ is isomorphic to the cyclic subgroup $S_6$ generated by the 6-cycle**

- $5 \Leftrightarrow \sigma = (1 2 3 4 5 6)$
- $4 \Leftrightarrow \sigma^2 = (1 3 5)(2 4 6)$  
- $6 \Leftrightarrow \sigma^3 = (1 4)(2 5)(3 6)$
- $2 \Leftrightarrow \sigma^4 = (1 5 3)(2 6 4)$
- $3 \Leftrightarrow \sigma^5 = (1 6 5 4 3 2)$
- $1 \Leftrightarrow \sigma^6 = e$

---

---

**Task 11**

For this task I used only the GAP script (poly_1.gap), which can be run with the command *gap -q poly_1.gap*, the output file is *poly_1_output.txt*

**Results:**
- **polynomial_F4**: $x^9+ x^8 + ax^6 + (a+1)x^5 + x^4 + ax^2 + (a+1)x + 1$
- **roots_F4**: [1]
- **polynomial_F7**: $5x^6 + 4x^5 + 3x^4 + 2x^3 + x^2 - 1$
- **roots_F7**: [4, 1], [1] - root of multiplicity 5

**Why:**
- $F_7$ is a prime field, all elements = ordinary numbers
- $F_4$ is a field extension, must be constructed using a polynomial

---

**Task 12**

For this task I also used only the GAP script (poly_2.gap), which can be run with the command *gap -q poly_2.gap*, the output file is *poly_2_output.txt*

**Results:**
- **poly_F5** = $x^5 + 2x^4 + x^3 - x + 3$
- **is_irreducible_F5** = true
- **poly_F9** = $x^4 + x^3 - x + 1$
- **is_irreducible_F9** = true

Both polynomials are irreducible

---

**Task 13**

For this task I decided to use only the SageMath script (poly_3.sage), which can be run with the command *sage poly_3.sage*, the output file is *poly_3_output.txt*

**Result:**
- **f** = $9x^7 + 8x^6 + 7x^5 + 6x^4 + 5x^3 + 4x^2 + 3x + 2$
- **g** = $5x^3 + 4x^2 + 3x + 2$
- **gcd** = $1$
- **u** = $3x^2 + 10x + 3$
- **v** = $10x^6 + 8x^4 + 9x^3 + 4x^2 + 3x + 3$
- **verification** = $1$

Therefore, $gcd(f, g) = u(x)f (x) + v(x)g(x) = 1$

---

**Task 14**

For this task I decided to use only the SageMath script (poly_4.sage), which can be run with the command *sage poly_4.sage*, the answer prints immediately. I also wrote tests for this.

**Result:**
- $f = 4x^2 + 3x + 2$
- $f^{-1}\ mod\ g = 9x^7 + 11x^6 + 10x^5 + 4x^3 + x^2 + 2x$

**Check:**
- $f * h\ mod\ g = 1$

Thats right

---

**Task 15**

For this task I use the SageMath script + python (poly_5.sage), which can be run with the command *sage poly_5.sage*.

**Results:**

q = 2, d = 2:
  - x^2 + x + 1
  
Count: 1

q = 2, d = 3:
  - x^3 + x + 1
  - x^3 + x^2 + 1

Count: 2

q = 2, d = 4:
  - x^4 + x + 1
  - x^4 + x^3 + 1
  - x^4 + x^3 + x^2 + x + 1

Count: 3

q = 3, d = 2:
  - x^2 + 1
  - x^2 + x + 2
  - x^2 + 2*x + 2

Count: 3

q = 3, d = 3:
  - x^3 + 2*x + 1
  - x^3 + 2*x + 2
  - ...
  - x^3 + 2*x^2 + x + 1
  - x^3 + 2*x^2 + 2*x + 2

Count: 8

q = 3, d = 4:
  - x^4 + x + 2
  - x^4 + 2*x + 2
  - ...
  - x^4 + 2*x^3 + x^2 + 2*x + 1
  - x^4 + 2*x^3 + 2*x^2 + x + 2

Count: 18

q = 5, d = 2:
  - x^2 + 2
  - x^2 + 3
  - ...
  - x^2 + 4*x + 1
  - x^2 + 4*x + 2

Count: 10

q = 5, d = 3:
  - x^3 + x + 1
  - x^3 + x + 4
  - ...
  - x^3 + 4*x^2 + 4*x + 2
  - x^3 + 4*x^2 + 4*x + 4

Count: 40

q = 5, d = 4:
  - x^4 + 2
  - x^4 + 3
  - ...
  - x^4 + 4*x^3 + 4*x^2 + 4*x + 1
  - x^4 + 4*x^3 + 4*x^2 + 4*x + 4

Count: 150

The polynomials that I missed due to their large number are in the file *poly_5_output.txt*

---